include/sys/async_notify.h

```diff
@@ -0,0 +1,346 @@
+ */
+#define ASYNC_NOTIFY_EXTENSION_MASK (~(BIT(ASYNC_NOTIFY_EXTENSION_POS) - 1U))
+
+/**
+ * @brief Generic signature used to notify of result completion by
+ * callback.
+ *
+ * Functions with this role may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-ok and not sleep.
+ *
+ * Parameters that should generally be passed to such functions include:
+ *
+ * * a pointer to a specific client request structure, i.e. the one
+ *   that contains the async_notify structure.
+ * * the result of the operation, either as passed to
+ *   async_notify_finalize() or extracted afterwards using
+ *   async_notify_fetch_result().  Expected values are
+ *   service-specific, but the value shall be non-negative if the
+ *   operation succeeded, and negative if the operation failed.
+ */
+typedef void (*async_notify_generic_callback)();
+
+/**
+ * @brief State associated with notification for an asynchronous
+ * operation.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. async_notify_init_signal()) to
+ * configure them.  Generally the structure is a member of a
+ * service-specific client structure, such as onoff_client.
+ *
+ * Control of the containing object transfers to the service provider
+ * when a pointer to the object is passed to a service function that
+ * is documented to take control of the object, such as
+ * onoff_service_request().  While the service provider controls the
+ * object the client must not change any object fields.  Control
+ * reverts to the client:
+ * * if the call to the service API returns an error;
+ * * if the call to the service API succeeds for a no-wait operation;
+ * * when operation completion is posted (signalled or callback
+ *   invoked).
+ *
+ * After control has reverted to the client the notify object must be
+ * reinitialized for the next operation.
+ *
+ * The content of this structure is not public API to clients: all
+ * configuration and inspection should be done with functions like
+ * async_notify_init_callback() and async_notify_fetch_result().
+ * However, services that use this structure may access certain
+ * fields directly.
+ */
+struct async_notify {
+	union method {
+		/* Pointer to signal used to notify client.
+		 *
+		 * The signal value corresponds to the res parameter
+		 * of async_notify_callback.
+		 */
+		struct k_poll_signal *signal;
+
+		/* Generic callback function for callback notification. */
+		async_notify_generic_callback callback;
+	} method;
+
+	/*
+	 * Flags recording information about the operation.
+	 *
+	 * Bits below ASYNC_NOTIFY_EXTENSION_POS must not by modified
+	 * by extensions.
+	 *
+	 * Bits at and above ASYNC_NOTIFY_EXTENSION_POS are available
+	 * for use by service extensions while the containing object
+	 * is managed by the service.  They are not for client use,
+	 * and are cleared before completion notfication.
```

Comment:
`async_notify_finalize` clears only the bits below ASYNC_NOTIFY_EXTENSION_POS. Was the intent to clear all of them?